lib/foreman_azure_rm/azure_sdk_adapter.rb

```diff
@@ -145,5 +173,23 @@ def stop_vm(rg_name, vm_name)
       compute_client.virtual_machines.power_off(rg_name, vm_name)
       compute_client.virtual_machines.deallocate(rg_name, vm_name)
     end
+
+    def self.gallery_caching(rg_name)
+      @gallery_caching ||= {}
+      @gallery_caching[rg_name] ||= {}
+    end
+
+    def actual_gallery_image_id(rg_name, image_id)
+      gallery_names = list_galleries.map(&:name)
+      gallery_names.each do |gallery|
+        gallery_image = list_gallery_images(rg_name, gallery).detect { |image| image.name == image_id }
+        return gallery_image.id
```

Comment:
if `detect` will return `nil`, you will have an `undefined method id for nil` exception. 
```suggestion
        return gallery_image&.id
```